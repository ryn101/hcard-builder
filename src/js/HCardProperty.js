/**
 * Configured types of hCard properties used as a lookup for value safety
 *
 * @type {{
 *    country: string,
 *    streetAddress: string,
 *    postalCode: string,
 *    locality: string,
 *    telephone: string,
 *    region: string,
 *    email: string
 *  }}
 */
const hCardPropertyType = {
    email: 'email',
    telephone: 'tel',
    streetAddress: 'street-address',
    locality: 'locality',
    region: 'region',
    postalCode: 'postal-code',
    country: 'country',
};

/**
 * Container for a HCard specific property
 */
class HCardProperty {
    /**
     * Constructor
     *
     * @param name The property name/type
     * @param value The value of the property
     */
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
}

export { hCardPropertyType as HCardPropertyType };
export default HCardProperty;
