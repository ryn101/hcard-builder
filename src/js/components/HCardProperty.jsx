import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 * Wraps up the logic needed to display a hCard property. The name of the property is added
 * as a class to the component and the title is set to the specified value.
 *
 * Additionally supports property nesting, and abbreviation declaration for use with the abbr tag.
 * If the abbreviation tag is provided then the as property is disregarded.
 * If child elements are provided then the value and abbreviation properties are disregarded.
 *
 * When requiring a HTML tag other than a span, the as property can be specified.
 */
function HCardProperty(props) {
    const {
        name,
        value,
        abbreviation,
        as,
        className,
        children,
    } = props;

    const ElementTag = as;
    let element;

    if (children || !abbreviation) {
        // If children are specified or no abbreviation value is specified then use the element tag
        element = (
            <ElementTag title={value} className={classNames(name, className)}>
                { children || value }
            </ElementTag>
        );
    } else {
        // If an abbreviation is specified then create an abbr tag instead
        element = (<abbr title={value} className={name}>{abbreviation}</abbr>);
    }

    return element;
}

HCardProperty.propTypes = {
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
    abbreviation: PropTypes.string,
    as: PropTypes.string,
    className: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.oneOfType([
            PropTypes.element,
            PropTypes.string,
        ])),
    ]),
};

HCardProperty.defaultProps = {
    value: '',
    as: 'span',
    className: '',
    children: null,
};

export default HCardProperty;
