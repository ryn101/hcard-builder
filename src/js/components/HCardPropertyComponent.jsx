import React from 'react';
import PropTypes from 'prop-types';
import HCardProperty from '../HCardProperty';

/**
 * Class storing details specific to a HCard property component
 */
class HCardPropertyComponent extends React.Component {
    /**
     * Renders a HCard property element specific to a singular property
     *
     * @param value The value of the property
     * @returns {*}
     */
    renderSingular = ({ value }) => (
        <>
            { value }
        </>
    );

    /**
     * Renders a Telephone HCard property element
     *
     * @param value The value of the property
     * @returns {*}
     */
    renderTelephone = ({ value }) => (
        <>
            <span className="type">Home</span>
            { value }
        </>
    );

    render() {
        const { hCardProperty } = this.props;

        let renderedElement;

        if (hCardProperty.name === 'tel') {
            renderedElement = this.renderTelephone(hCardProperty);
        } else {
            renderedElement = this.renderSingular(hCardProperty);
        }

        return (
            <span title={hCardProperty.value} className={hCardProperty.name}>
                { renderedElement }
            </span>
        );
    }
}

HCardPropertyComponent.propTypes = {
    hCardProperty: PropTypes.instanceOf(HCardProperty).isRequired,
};

export default HCardPropertyComponent;
