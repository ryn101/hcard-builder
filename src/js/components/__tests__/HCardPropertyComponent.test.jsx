import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HCardProperty from '../../HCardProperty';
import HCardPropertyComponent from '../HCardPropertyComponent';

configure({ adapter: new Adapter() });

const expectBasicDetailsToBePopluated = (hCardPropertyNode, hCardProperty) => {
    expect(hCardPropertyNode.prop('title')).toEqual(hCardProperty.value);
    expect(hCardPropertyNode.hasClass(hCardProperty.name)).toBeTruthy();
};

it('renders the hCard property without crashing', () => {
    const hCardProp = new HCardProperty('prop1', 'value1');
    const div = document.createElement('div');
    ReactDOM.render(<HCardPropertyComponent hCardProperty={hCardProp} />, div);
    ReactDOM.unmountComponentAtNode(div);
});

describe('once the hcard detail row has rendered', () => {
    let hCardPropertyNode;
    const hCardProperty = new HCardProperty();

    beforeEach(() => {
        hCardPropertyNode = shallow(<HCardPropertyComponent hCardProperty={hCardProperty} />);
    });

    it('renders with state correctly', () => {
        expect(hCardPropertyNode).toMatchSnapshot();
    });

    it('renders a singular property with the required details', () => {
        hCardProperty.name = 'Prop1';
        hCardProperty.value = 'Value1';
        hCardPropertyNode.setProps({ hCardProperty });
        hCardPropertyNode.update();

        expectBasicDetailsToBePopluated(hCardPropertyNode, hCardProperty);
        expect(hCardPropertyNode.text()).toEqual(hCardProperty.value);
    });

    it('renders a telephone property with the required details', () => {
        hCardProperty.name = 'tel';
        hCardProperty.value = '92825925';
        hCardPropertyNode.setProps({ hCardProperty });
        hCardPropertyNode.update();

        expectBasicDetailsToBePopluated(hCardPropertyNode, hCardProperty);
        const typeNode = hCardPropertyNode.find('.type');
        expect(typeNode).toHaveLength(1);
        expect(hCardPropertyNode.text()).toContain(hCardProperty.value);
    });
});
