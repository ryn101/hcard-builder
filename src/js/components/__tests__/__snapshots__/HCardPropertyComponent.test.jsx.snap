// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once the hcard detail row has rendered renders with state correctly 1`] = `<span />`;
